### ISSUE TITLE
Develop Algorithm for Collecting and Flattening Weather Data 

### ISSUE THREAD

**Issue by rectheworld** [DataKind-DC/audubon-cbc#28]
NOTE: We will be using 2019 matched station data to develop the algorithm because the 2019 data has the best station matches. For the real data collection we will be using data from 1901 through 2019 (This work will begin after the data engineers rematch to closest stations) PLEASE FILTER TO 2019 WHEN DEVELOPING

Inside the google project in big Query and GCS is a table/file called usa_cleaned_matched. This data has DRAFT matches for each of the cbc circles matched to their closest station. We have found that only the 2019 stations are accurate, so those should be used for developing this algorithm. 

GOAL: For each Station for a given day (the count_date), pull the weather data from the official weather source. 

SUMMARY:

INPUTS: 
- Station_id (from usa_cleaned_matched)
- count_date (from usa_cleaned_matched)

OUTPUTS: 
Weather data formatted like this:
Station ID | PREC | TEMP HIGH | TEMP LOW
-- | -- | -- | --
ABCD1234 | 2 | 97 | 67

# Where were getting the weather data
Official Data Source: https://console.cloud.google.com/marketplace/details/noaa-public/ghcn-d?filter=solution-type:dataset&id=9d500d1d-fda4-4413-a789-d8786fd6592e&pli=1

This Google BigQuery Database is a collection of daily weather measurements form NOAA. Each Year of data is divided into its own table, ie 2019 is gcnh_2019. 

When the data is returned to the user it looks like this:

Station ID | Data Type | Value
-- | -- | --
ABCD1234 | PREC | 2
ABCD1235 | TEMP HIGH | 97
ABCD1236 | TEMP LOW | 67

For each station we want it to look like this:

Station ID | PREC | TEMP HIGH | TEMP LOW
-- | -- | -- | --
ABCD1234 | 2 | 97 | 67

The goal is to be able to merge this flattened weather data with our usa_cleaned_matched data.

**Comment by rectheworld** [comment]
# Demo Query to Flatten the BigQuery weather data:

""""
SELECT DISTINCT y.id, y.date, 
(SELECT value FROM `bigquery-public-data.ghcn_d.ghcnd_2019` x WHERE x.element = 'TMIN' AND x.date = y.date AND x.id = y.id) as t_min,
(SELECT value FROM `bigquery-public-data.ghcn_d.ghcnd_2019` x WHERE x.element = 'TMAX' AND x.date = y.date AND x.id = y.id) as t_max
FROM `bigquery-public-data.ghcn_d.ghcnd_2019` y LIMIT 1000
"""

**Comment by rectheworld** [comment]
From Francisco Vannini 

This is a python script that grabs the ghcnd data corresponding to the years 2010 up to 2019 (for the station at Washington Dulles Airport) and appends it to a dataframe. It queries the BigQuery API so you are going to have to setup the credentials as stated here (https://cloud.google.com/bigquery/docs/authentication/) and assign the right role priviledge for the corresponding user

```
PROJECT='fjvr-testing'  # CHANGE THIS
from google.cloud import bigquery
from google.cloud.bigquery.job import LoadJobConfig, WriteDisposition, CreateDisposition
import pandas as pd
df = pd.DataFrame(columns=['date','id','name','precipitation_value','state','temp_max_value','temp_min_value'])
bq = bigquery.Client(project=PROJECT)
initial_year = 2010
# list tables in dataset
tables = bq.list_tables("bigquery-public-data.ghcn_d")
for table in tables:
    try:
        if int(table.table_id[-4:]) >= initial_year:
            query = """
            SELECT DISTINCT
              base.id, 
              base.date,
              stations.name,
              stations.state,
              temp_min.value as temp_min_value,
              temp_max.value as temp_max_value,
              precipitation.value as precipitation_value
            FROM {} base
            INNER JOIN {} temp_min ON base.id = temp_min.id AND base.date = temp_min.date
            INNER JOIN {} temp_max ON base.id = temp_max.id AND base.date = temp_max.date
            INNER JOIN {} precipitation ON base.id = precipitation.id AND base.date = precipitation.date
            INNER JOIN `bigquery-public-data`.ghcn_d.ghcnd_stations stations ON base.id = stations.id
            WHERE temp_min.element = 'TMIN' AND temp_max.element = 'TMAX' AND precipitation.element = 'PRCP'
            AND stations.state = 'VA' AND stations.id = 'USW00013743'
            ORDER BY base.id, base.date
            """
            parameter = "`bigquery-public-data`.ghcn_d." + table.table_id
            print(parameter)
            query = query.format(parameter,parameter,parameter,parameter)
            df_temp = bq.query(query).to_dataframe()
            df = df.append(df_temp)
    except:
        pass
df
```

**Comment by Frankie-Figz** [comment]
Here is an update script that saves the query as an csv. It also has comments.

```
import pandas as pd # Read in unfinished dataset
from google.cloud import bigquery
from google.cloud.bigquery.job import LoadJobConfig, WriteDisposition, CreateDisposition
import pandas as pd# Change this to the PATH of your cbc_cleaned_usa_statid station
df_shell = pd.read_csv('C:\\Users\\Frankie-Figz\\Downloads\\cbc_cleaned_usa_statid.csv')# Find the required dates and station ids 
station_list = df_shell.loc[:, 'closest_station_id'].unique()
date_list = df_shell.loc[:,'count_date'].unique()# Format the dates and stations ids as to be acceptable to the query that will later be formed
string_parameter_stations  = ""
seperator = ' \',\' '
string_parameter_stations = seperator.join(station_list)
string_parameter_stations = "(\'" + string_parameter + "\')"
string_parameter_stations = string_parameter.replace(" ", "")string_parameter_dates = ""
string_parameter_dates = seperator.join(date_list)
string_parameter_dates = "(\'" + string_parameter_dates + "\')"
string_parameter_dates = string_parameter_dates.replace(" ", "")# This is the name of your google project. It is related to the key you use to register with google API services
# For more information see : https://cloud.google.com/bigquery/docs/authentication/
PROJECT='fjvr-testing'  # CHANGE THIS TO YOUR PROJECT # Create the date frame that will be used to accumulate the results from each year
df = pd.DataFrame(columns=['date','id','name','precipitation_value','state','temp_max_value','temp_min_value'])# Instantiate a big query object
bq = bigquery.Client(project=PROJECT)# The initial date and final date that the query will use to retrieve data
initial_year = 2018
final_year = 2019#A table for every year of observations is used in the NOAA dataset.
# For each year we query the dataset and thus have to reformulate the table name# list tables in the big query GHCN dataset
tables = bq.list_tables("bigquery-public-data.ghcn_d")# Iterates over all tables in the bigquery-public-data.ghcn_d datasets and only takes the once that have a year as the last 4 digits.
for table in tables:
    try:
        if int(table.table_id[-4:]) >= initial_year and int(table.table_id[-4:]) <= final_year:
            # Constructs a query to flatten data. Here we are doing a SELF-JOIN (see examples online if neccesary) to extract the attributes of interest
            query = """
            SELECT DISTINCT
              base.id, 
              base.date,
              stations.name,
              stations.state,
              temp_min.value as temp_min_value,
              temp_max.value as temp_max_value,
              precipitation.value as precipitation_value            FROM {} base
            INNER JOIN {} temp_min ON base.id = temp_min.id AND base.date = temp_min.date
            INNER JOIN {} temp_max ON base.id = temp_max.id AND base.date = temp_max.date
            INNER JOIN {} precipitation ON base.id = precipitation.id AND base.date = precipitation.date
            INNER JOIN `bigquery-public-data`.ghcn_d.ghcnd_stations stations ON base.id = stations.id            WHERE temp_min.element = 'TMIN' AND temp_max.element = 'TMAX' AND precipitation.element = 'PRCP' AND stations.id IN {} AND base.date IN {}            ORDER BY base.id, base.date
            """
            # This parameter is determined baased on the current year we are iterating over
            parameter = "`bigquery-public-data`.ghcn_d." + table.table_id            # The format function replaces instances of {} with the corresponding sequential input
            query = query.format(parameter,parameter,parameter,parameter,string_parameter_stations, string_parameter_dates)            # Queries BigQuery public data set and creates a new dataframe object
            df_temp = bq.query(query).to_dataframe()            # Appends the dataframe object to accumulated data frame
            df = df.append(df_temp)
    except:
        pass# Change the PATH relative to you
df.to_csv(r'C:\\Users\\Frankie-Figz\\Documents\\code-space\\audubon-cbc\\datasets\\flatten_data_from_' + str(initial_year) + '_to_' + str(final_year) + '.csv')
```

**Comment by rectheworld** [comment]
Committing this script to the Gihub so we can close

**Comment by Frankie-Figz** [comment]
Hi Ren, could this issue be closed or are their additional tasks required?

**Comment by rectheworld** [comment]
Yes this ones can be closed. 

Closed with pull request #34
